pkg/utils/map.go

```diff
@@ -14,6 +19,54 @@ func CleanUpInterfaceMap(in map[interface{}]interface{}) map[string]interface{}
 	return result
 }
 
+func RemoveNulls(m map[string]interface{}) {
+	val := reflect.ValueOf(m)
+	for _, e := range val.MapKeys() {
+		v := val.MapIndex(e)
+		if v.IsNil() {
+			delete(m, e.String())
+			continue
+		}
+
+		t, ok := v.Interface().(map[string]interface{})
+		if ok {
+			RemoveNulls(t)
+		}
+	}
+}
+
+func PatchInterfaceMap(defaultValues, values map[string]map[string]interface{}) (map[string]map[string]interface{}, error) {
```

Comment:
Why are we using this approach here?

Reply:
this is for migration from values.yaml to default-values.yaml. We make patch from those two files to find the only changes (value changes, new fields) and then we override values.yaml with those changes

Reply:
right now the `values.yaml` contains only changes to override

Reply:
in that case, if this is just for migration purposes, we should have some conditional code to ensure we only do it when the default-values.yaml file has not yet been created.

Reply:
Generally i think this should be programmed as defensively as possible

Reply:
The way it runs now actually will allow us to retroactively clean up the values files, which some apps can definitely benefit from. The migration also only happens for keys that have the same value that we template. So for example, if a user changes the OIDC client ID and secret in the `values.yaml` file that will not get migrated to the `default-values.yaml`.

Reply:
Will we ever need to do this beyond the one initial time the new default-values.yaml file is created though?  If not, we should disable the code path from then on, by checking if it exists and ignoring.

Reply:
At the moment there are a number of apps we could do this for because we were lacking this functionality previously. One example is Argo-CD or Plural itself. For those applications we could push a chart version including the old templated values to get those into the `default-values.yaml` and then another update to clean them up.

Reply:
I've also checked it with arrays and if people customize an array that we template it is left in the `values.yaml` properly, overriding what we template as would be expected.